File: typescript/inference.ts
/**
 * IPIP-BFFM Sparse Quantile Model — TypeScript inference module.
 *
 * Loads a single merged ONNX model and provides prediction from either a
 * Record of item responses or a raw Float32Array.
 */

import { readFileSync } from "node:fs";
import { join, resolve } from "node:path";
import * as ort from "onnxruntime-node";

const DOMAINS = ["ext", "agr", "csn", "est", "opn"] as const;
const QUANTILES = ["q05", "q50", "q95"] as const;

type Domain = (typeof DOMAINS)[number];
type Quantile = (typeof QUANTILES)[number];

export interface DomainResult {
  raw: { q05: number; q50: number; q95: number };
  percentile: { q05: number; q50: number; q95: number };
}

export type PredictionResult = Record<Domain, DomainResult>;

interface CalibrationDomain {
  observed_coverage: number;
  scale_factor: number;
}

interface ModelConfig {
  domains: string[];
  model_file: string;
  outputs: string[];
  input: { feature_names: string[] };
  norms: Record<string, { mean: number; sd: number }>;
  calibration?: Record<string, Record<string, CalibrationDomain>>;
}

/**
 * Abramowitz & Stegun (1964) approximation of the standard normal CDF.
 * Same implementation as packages/web/lib/utils.ts.
 */
function standardNormalCDF(z: number): number {
  if (!Number.isFinite(z)) return Number.NaN;

  const absZ = Math.abs(z);
  const t = 1 / (1 + 0.2316419 * absZ);
  const d = Math.exp(-0.5 * absZ * absZ) / Math.sqrt(2 * Math.PI);

  const poly =
    t *
    (0.31938153 +
      t *
        (-0.356563782 +
          t * (1.781477937 + t * (-1.821255978 + t * 1.330274429))));

  const cnd = 1 - d * poly;
  const p = z >= 0 ? cnd : 1 - cnd;
  return Math.min(1, Math.max(0, p));
}

export class IPIPBFFMPredictor {
  private session!: ort.InferenceSession;
  private config: ModelConfig;
  private featureIndex: Map<string, number>;
  private outputNames: string[];

  private constructor(config: ModelConfig) {
    this.config = config;
    this.outputNames = config.outputs;
    this.featureIndex = new Map(
      config.input.feature_names.map((name, i) => [name, i])
    );
  }

  /**
   * Create a predictor by loading config and the merged ONNX session.
   *
   * @param modelDir Path to the directory containing model.onnx and config.json.
   *                 Defaults to the parent of this file's directory.
   */
  static async create(modelDir?: string): Promise<IPIPBFFMPredictor> {
    const dir = modelDir ?? resolve(import.meta.dirname, "..", "output", "reference");
    const configPath = join(dir, "config.json");
    const config: ModelConfig = JSON.parse(
      readFileSync(configPath, "utf-8")
    );

    const predictor = new IPIPBFFMPredictor(config);
    predictor.session = await ort.InferenceSession.create(
      join(dir, config.model_file)
    );

    return predictor;
  }

  /**
   * Predict from a Record mapping item IDs to response values (1-5).
   * Missing items are treated as NaN (unanswered).
   *
   * @throws {Error} If any keys in `responses` are not valid item IDs.
   */
  async predict(
    responses: Record<string, number>
  ): Promise<PredictionResult> {
    const unknown = Object.keys(responses)
      .filter((k) => !this.featureIndex.has(k))
      .sort();
    if (unknown.length > 0) {
      const valid = this.config.input.feature_names.join(", ");
      throw new Error(
        `Unrecognized item IDs: ${unknown.join(", ")}. Valid IDs: ${valid}`
      );
    }

    const bad: Record<string, number> = {};
    for (const [k, v] of Object.entries(responses)) {
      if (!Number.isFinite(v) || v < 1 || v > 5) {
        bad[k] = v;
      }
    }
    if (Object.keys(bad).length > 0) {
      throw new Error(
        `Response values must be finite numbers in [1, 5]. Got: ${JSON.stringify(bad)}`
      );
    }

    const n = this.config.input.feature_names.length;
    const arr = new Float32Array(n).fill(NaN);
    for (const [itemId, value] of Object.entries(responses)) {
      const idx = this.featureIndex.get(itemId);
      if (idx !== undefined) {
        arr[idx] = value;
      }
    }
    return this.predictArray(arr);
  }

  private calibrationRegime(responses: Float32Array): string {
    let nAnswered = 0;
    for (const v of responses) {
      if (!Number.isNaN(v)) nAnswered++;
    }
    return nAnswered >= 50 ? "full_50" : "sparse_20_balanced";
  }

  /**
   * Predict from a raw Float32Array of length 50.
   * NaN for unanswered items.
   *
   * @throws {Error} If `responses` does not have the expected length.
   */
  async predictArray(responses: Float32Array): Promise<PredictionResult> {
    const n = this.config.input.feature_names.length;
    if (responses.length !== n) {
      throw new Error(
        `Expected Float32Array of length ${n}, got ${responses.length}`
      );
    }

    const tensor = new ort.Tensor("float32", responses, [1, n]);
    const output = await this.session.run({ input: tensor });

    const regime = this.calibrationRegime(responses);
    const regimeCal = this.config.calibration?.[regime] ?? {};

    const results: Partial<PredictionResult> = {};

    for (const domain of DOMAINS) {
      const rawScores: Record<string, number> = {};
      for (const q of QUANTILES) {
        const key = `${domain}_${q}`;
        rawScores[q] = (output[key].data as Float32Array)[0];
      }

      // Convert raw scores to percentiles
      const norms = this.config.norms[domain];
      const pct: Record<string, number> = {};
      for (const q of QUANTILES) {
        const z = (rawScores[q] - norms.mean) / norms.sd;
        pct[q] = standardNormalCDF(z) * 100;
      }

      // Sort percentiles so q05 <= q50 <= q95
      const pctVals = [pct.q05, pct.q50, pct.q95].sort((a, b) => a - b);
      pct.q05 = pctVals[0];
      pct.q50 = pctVals[1];
      pct.q95 = pctVals[2];

      const scale = regimeCal[domain]?.scale_factor ?? 1.0;
      if (scale !== 1.0) {
        const halfWidth = 0.5 * (pct.q95 - pct.q05) * scale;
        pct.q05 = Math.min(100, Math.max(0, pct.q50 - halfWidth));
        pct.q95 = Math.min(100, Math.max(0, pct.q50 + halfWidth));
      }

      results[domain] = {
        raw: {
          q05: Math.round(rawScores.q05 * 10000) / 10000,
          q50: Math.round(rawScores.q50 * 10000) / 10000,
          q95: Math.round(rawScores.q95 * 10000) / 10000,
        },
        percentile: {
          q05: Math.round(pct.q05 * 10) / 10,
          q50: Math.round(pct.q50 * 10) / 10,
          q95: Math.round(pct.q95 * 10) / 10,
        },
      };
    }

    return results as PredictionResult;
  }

  /** Release the ONNX session. */
  dispose(): void {
    this.session?.release();
  }
}

File: python/inference.py
"""
IPIP-BFFM Sparse Quantile Model — Python inference module.

Loads a single merged ONNX model and provides prediction from either a dict
of item responses or a raw numpy array.
"""

from __future__ import annotations

import json
from pathlib import Path

import numpy as np
import onnxruntime as ort
from scipy.stats import norm

DOMAINS = ["ext", "agr", "csn", "est", "opn"]
QUANTILES = ["q05", "q50", "q95"]


class IPIPBFFMPredictor:
    """Predict Big Five personality scores from IPIP-BFFM item responses."""

    def __init__(self, model_dir: str | Path | None = None) -> None:
        """Load config.json and the merged ONNX session.

        Args:
            model_dir: Path to the directory containing model.onnx and config.json.
                       Defaults to the parent of this file's directory.
        """
        if model_dir is None:
            model_dir = Path(__file__).resolve().parent.parent / "output" / "reference"
        self.model_dir = Path(model_dir)

        with open(self.model_dir / "config.json") as f:
            self.config = json.load(f)

        self.feature_names: list[str] = self.config["input"]["feature_names"]
        self._feature_index = {
            name: i for i, name in enumerate(self.feature_names)
        }
        self.norms: dict = self.config["norms"]

        # Load single merged ONNX session
        model_path = self.model_dir / self.config["model_file"]
        self._session = ort.InferenceSession(str(model_path))
        self._output_names: list[str] = self.config["outputs"]

    def predict(self, responses: dict[str, float]) -> dict:
        """Predict from a dict mapping item IDs to response values (1-5).

        Missing items are treated as NaN (unanswered).

        Args:
            responses: e.g. {"ext3": 4.0, "agr1": 3.0, ...}

        Returns:
            Per-domain results with raw scores and percentiles for each quantile.

        Raises:
            ValueError: If any keys in *responses* are not valid item IDs.
        """
        unknown = sorted(set(responses) - set(self._feature_index))
        if unknown:
            valid = ", ".join(self.feature_names)
            raise ValueError(
                f"Unrecognized item IDs: {', '.join(unknown)}. "
                f"Valid IDs: {valid}"
            )

        bad = {
            k: v for k, v in responses.items()
            if not (isinstance(v, (int, float)) and np.isfinite(v) and 1 <= v <= 5)
        }
        if bad:
            raise ValueError(
                f"Response values must be finite numbers in [1, 5]. "
                f"Got: {bad}"
            )

        arr = np.full((1, len(self.feature_names)), np.nan, dtype=np.float32)
        for item_id, value in responses.items():
            idx = self._feature_index.get(item_id)
            if idx is not None:
                arr[0, idx] = value
        return self.predict_array(arr)

    def _calibration_regime(self, responses: np.ndarray) -> str:
        n_answered = int(np.sum(~np.isnan(responses)))
        return "full_50" if n_answered >= 50 else "sparse_20_balanced"

    def predict_array(self, responses: np.ndarray) -> dict:
        """Predict from a raw [1, 50] float32 array.

        Args:
            responses: Shape (1, 50) float32 array. NaN for unanswered items.

        Returns:
            Per-domain results with raw scores and percentiles for each quantile.

        Raises:
            ValueError: If *responses* does not have the expected shape.
        """
        expected_shape = (1, len(self.feature_names))
        if responses.shape != expected_shape:
            raise ValueError(
                f"Expected array with shape {expected_shape}, "
                f"got {responses.shape}"
            )

        # Single session.run() for all 15 outputs
        raw_outputs = self._session.run(
            self._output_names, {"input": responses}
        )
        raw_dict = {
            name: float(output.flatten()[0])
            for name, output in zip(self._output_names, raw_outputs)
        }

        calibration = self.config.get("calibration", {})
        regime = self._calibration_regime(responses[0])
        regime_cal = calibration.get(regime, {})

        results = {}
        for domain in DOMAINS:
            raw_scores = {}
            for q in QUANTILES:
                key = f"{domain}_{q}"
                raw_scores[q] = raw_dict[key]

            # Convert raw scores to percentiles
            n = self.norms[domain]
            pct = {}
            for q in QUANTILES:
                z = (raw_scores[q] - n["mean"]) / n["sd"]
                pct[q] = float(norm.cdf(z) * 100)

            # Sort percentiles so q05 <= q50 <= q95
            pct_vals = sorted([pct["q05"], pct["q50"], pct["q95"]])
            pct["q05"], pct["q50"], pct["q95"] = pct_vals

            scale = float(regime_cal.get(domain, {}).get("scale_factor", 1.0))
            if scale != 1.0:
                half_width = 0.5 * (pct["q95"] - pct["q05"]) * scale
                pct["q05"] = float(np.clip(pct["q50"] - half_width, 0.0, 100.0))
                pct["q95"] = float(np.clip(pct["q50"] + half_width, 0.0, 100.0))

            results[domain] = {
                "raw": {q: round(raw_scores[q], 4) for q in QUANTILES},
                "percentile": {q: round(pct[q], 1) for q in QUANTILES},
            }
        return results
